src/api_types.jl

```diff
@@ -0,0 +1,329 @@
+## C types
+const C_time_t = Int
+
+## HDF5 types and constants
+const haddr_t  = UInt64
+const herr_t   = Cint
+const hid_t    = Int64
+const hsize_t  = UInt64
+const hssize_t = Int64
+const htri_t   = Cint   # pseudo-boolean (negative if error)
+
+# For VLEN
+struct hvl_t
+    len::Csize_t
+    p::Ptr{Cvoid}
+end
+const HVL_SIZE = sizeof(hvl_t)
+
+# Reference types
+struct hobj_ref_t
+    buf::NTuple{8,UInt8} # H5R_OBJ_REF_BUF_SIZE bytes
+end
+struct hdset_reg_ref_t
+    buf::NTuple{12,UInt8} # H5R_DSET_REG_REF_BUF_SIZE bytes
+end
+const HOBJ_REF_T_NULL = hobj_ref_t(ntuple(_ -> 0x0, Val(8)))
+const HDSET_REG_REF_T_NULL = hdset_reg_ref_t(ntuple(_ -> 0x0, Val(12)))
+
+#primitive type hobj_ref_t 64 end # size is H5R_OBJ_REF_BUF_SIZE bytes
+#primitive type hdset_reg_ref_t 96 end # size is H5R_DSET_REG_REF_BUF_SIZE bytes
+#const HOBJ_REF_T_NULL = reinterpret(type64_t, UInt64(0))
+#const HDSET_REG_REF_T_NULL = Core.Intrinsics.zext_int(hdset_reg_ref_t, UInt(0))
+
+# TODO: when upgraded to using newer HDF5 v1.12 reference API, can replace both with just
+#=
+struct H5R_ref_t
+    buf::NTuple{64,UInt8} # H5R_REF_BUF_SIZE bytes
+end
```

Comment:
I remember looking at this a little bit before but then abandoning it because we weren't up to v1.12 yet. I remember being a little concerned about alignment issues since the c code uses a union to enforce 64 byte alignment which I wasn't sure how to mimic in julia.

```
typedef struct {
    union {
        uint8_t __data[H5R_REF_BUF_SIZE];       /* opaque data */
        int64_t align;                          /* ensures alignment */
    } u;
} H5R_ref_t;
```

Do you know if this is a real issue or does everything just work?

Reply:
Good catch — I originally had it has 8 `Int64`s, but changed it later on having forgotten this point.

From the [Developer Docs](https://docs.julialang.org/en/v1/devdocs/object/#Object-allocation), you'll get pointer alignment on GC allocated resources, so it'll be correct by default on 64-bit systems, but 32-bit systems could be a problem. Also, my reading of that statement is that that only applies to GC resources, so I'm not sure[*] that carries forward to stack variables if the compiler can elide everything. To take a stab at empirically determining what happens, I've made up [this simple trial](https://gist.github.com/jmert/e60e6f0b49eed89b527a1cabddc4d23b) — I think it shows that you do end up with only 1-byte alignment if defined with `UInt8`s while `UInt64`s get you the desired 8-byte alignment.

Therefore I'll just change the struct definition to using `NTuple{8,UInt64}` instead.

_[*] (My maybe incorrect understanding of CPU architectures, though, is that x86 is fine loading unaligned 8-byte values, so optimizing at `Ref` which under GC would be pointer-aligned to an unaligned stack allocation is OK, but presumably Julia/LLVM would keep a larger alignment on architectures where alignment matters to maintain the semantics?? That doesn't change the conclusion that defining with `UInt64`s is still what you want to avoid issues on 32-bit systems ... just something I'm curious about.)_